### DISCUSSION TITLE
Using lifespan to create API in a gunicorn context increases required resources to boot up

### DISCUSSION THREAD

**Discussion by cristianbercu** [fastapi/fastapi#11227] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
@asynccontextmanager
async def lifespan(app: FastAPI) -> Any:
    # other before operations here ..
    # ...
    if connect_to_db:
        create_db(cfg, session_scopefunc=get_request_id)

    yield
    # tear down database connection
    logger.info("Application shutdown - destroy db..")
    destroy_db()


def create_api() -> FastAPI:
    return FastAPI(
        lifespan=lifespan
    )
```

### Description

I am using gunicorn to launch 4 workers as per config, and I am running a very lite 0.25 vCPU 0.5GB RAM container in AWS ECS Fargate.

Gunicorn config:
```python
bind = "0.0.0.0:8000"
workers = 4
worker_class = "uvicorn.workers.UvicornWorker"
keepalive = 5  # to make sure LB has time to forward request
```

I was using FastAPI `0.99.1` and I wanted to upgrade to the latest `1.110.1` but that forced me to start using the `lifespan` event for my before and after API launch. In the older version, I was using the `on_event("startup")` and `on_event("shutdown")` but now they are flagged as deprecated.

I changed my code to use `lifespan` with `yield` but this is requesting more CPU power to bootup FastAPI instances and gunicorn throws an out of memory exception "Perhaps out of memory?". This forced me to upgrade to 0.5 vCPU and 1GB RAM container instance types and spend more $. FastAPI should be lite, why is the new version consuming more CPU? 

### Operating System

Linux

### Operating System Details

Running in a Docker container on ECS.

### FastAPI Version

1.110.0

### Pydantic Version

2.6.3

### Python Version

3.10.12

**Comment by Kludex** [comment]
It shouldn't. Do you have an MRE?

**Reply by cristianbercu** [reply]
I am not sure what "MRE" stands for. If you are referring to a middleware, than yes. I have a small wrapper that sets some context to each request such as an unique identifier to enrich application logs. This is also pretty lite and unlikely to cause issues.

**Reply by n8sty** [reply]
MRE = minimal reproducible example

**Comment by n8sty** [comment, marked as answer]
Using 4 workers is high relative to gunicorn's suggestion of `(2 x $num_cores) + 1` (https://docs.gunicorn.org/en/stable/design.html#how-many-workers). I don't know your exact AWS setup, but when using orchestrators like ECS it oftentimes makes sense to scale the number of task instances running for a service rather than scaling the number of workers within a single instance. I imagine you were running right at the limit of available resource and some subtle change tipped you over the edge.

**Reply by cristianbercu** [reply]
Thanks for your answer @n8sty. I did indeed optimised (perhaps too much) my resource consumption to not use more than what I need which I think is generally a good practice. My container/instance setup is as light as it gets, nothing fancy about it. I think this is indeed related to gunicorn and the config. I should spawn 1-2 workers only based on their recommended rule which is fair. However, in my previous configuration I've set gunicorn to spawn only 1 worker but that was throwing the same error so I concluded that fastapi is using more resources at start-up in the new version as if I switch to the older version, I can spawn 4 workers on minimal hardware.

**Comment by luhclalala** [comment]
Sir, I encountered an issue while using Gunicorn with FastAPI. I attempted to start predefined code using the following approach. In the Gunicorn startup configuration, I used the `--preload` parameter, hoping to preload some configuration files or AI models in the main process so that worker processes could directly inherit them. However, this loading does not occur when FastAPI starts running. Instead, it only begins when I access the database-related API endpoints in FastAPI. General endpoints do not trigger the loading process. Could you please help me?

```python
from api.routers.preload import preload_manager
@asynccontextmanager
async def lifespan(app: FastAPI):
    await preload_manager.preload_all()
    yield
    await preload_manager.cleanup()
app = FastAPI(lifespan=lifespan)
app.include_router(questions.router)
```
